vars/yamlSubstituteVariables.groovy

```diff
@@ -0,0 +1,246 @@
+ * @param manifestNode - the manifest YAML to replace variables in.
+ * @param variablesData - the variables values.
+ * @param context - an execution context that can be used to query if any variables were replaced.
+ * @return a YAML object graph which has all variables replaced.
+ */
+private Object substitute(Object manifestNode, Object variablesData, ExecutionContext context) {
+    Map<String, Object> variableSubstitutes = getVariableSubstitutes(variablesData)
+
+    if (containsVariableReferences(manifestNode)) {
+
+        Object complexResult = null
+        String stringNode = manifestNode as String
+        Map<String, String> referencedVariables = getReferencedVariables(stringNode)
+        referencedVariables.each { referencedVariable ->
+            String referenceToReplace = referencedVariable.getKey()
+            String referenceName = referencedVariable.getValue()
+            Object substitute = variableSubstitutes.get(referenceName)
+
+            if (null == substitute) {
+                echo  "[YamlSubstituteVariables] ERROR - Found variable reference ${referenceToReplace} in input Yaml but no variable value to replace it with."
+                echo  "[YamlSubstituteVariables] ERROR - Leaving it unresolved. Check your variables Yaml data and make sure the variable is properly declared."
+                echo  "[YamlSubstituteVariables] ERROR - Unresolved variables may lead to follow-up problems (e.g. during a CF deployment). Failing this build."
+                error "[YamlSubstituteVariables] Not all variables could be resolved."
+            }
+
+            echo "Replacing: ${referenceToReplace} with ${substitute}"
+
+            if(isSingleVariableReference(stringNode)) {
+                logger.debug("Node ${stringNode} is SINGLE variable reference. Substitute type is: ${substitute.getClass().getName()}")
+                // if the string node we need to do replacements for is
+                // a reference to a single variable, i.e. should be replaced
+                // entirely with the variable value, we replace the entire node
+                // with the variable's value (which can possibly be a complex type).
+                complexResult = substitute
+            }
+            else {
+                logger.debug("Node ${stringNode} is multi-variable reference or contains additional string constants. Substitute type is: ${substitute.getClass().getName()}")
+                // if the string node we need to do replacements for contains various
+                // variable references or a variable reference and constant string additions
+                // we do a string replacement of the variables inside the node.
+                String regex = "\\(\\(${referenceName}\\)\\)"
+                stringNode = stringNode.replaceAll(regex, substitute as String)
+            }
+        }
+        context?.noVariablesReplaced = false  // remember that variables were found in the YAML file that have been replaced.
```

Comment:
When I call the step without context I get: `Cannot set property 'noVariablesReplaced' on null object`
So the context is required, we should not fail here like this in case of a required context, but above after preparing the configuration. To my knowledge `?.` does not work for setting values in case case the part before is `null`.

Reply:
That's weird, since I tested this and it works on my machine. Must be a quirk of the Jenkins Groovy environment. 

Fixed.